File: go/mechanisms/evm/deferred/server/scheme.go
// Package server provides the deferred EVM payment scheme for x402 sellers.
//
// From the seller's perspective, the deferred scheme behaves identically to
// "exact" for building payment requirements — the difference is handled
// entirely by the Facilitator during verify/settle. The Facilitator converts
// session-key signatures to EOA signatures and batches settlements on-chain.
package server

import (
	"context"

	x402 "github.com/okx/payments/go"
	"github.com/okx/payments/go/mechanisms/evm"
	exact "github.com/okx/payments/go/mechanisms/evm/exact/server"
	"github.com/okx/payments/go/types"
)

// AggrDeferredEvmScheme implements SchemeNetworkServer for deferred EVM payments.
// It delegates all logic to ExactEvmScheme, differing only in scheme name.
type AggrDeferredEvmScheme struct {
	exact *exact.ExactEvmScheme
}

// NewAggrDeferredEvmScheme creates a new AggrDeferredEvmScheme.
func NewAggrDeferredEvmScheme() *AggrDeferredEvmScheme {
	return &AggrDeferredEvmScheme{
		exact: exact.NewExactEvmScheme(),
	}
}

// Scheme returns "aggr_deferred".
func (s *AggrDeferredEvmScheme) Scheme() string {
	return evm.SchemeAggrDeferred
}

// ParsePrice delegates to the exact scheme's ParsePrice.
func (s *AggrDeferredEvmScheme) ParsePrice(price x402.Price, network x402.Network) (x402.AssetAmount, error) {
	return s.exact.ParsePrice(price, network)
}

// EnhancePaymentRequirements delegates to the exact scheme's EnhancePaymentRequirements.
func (s *AggrDeferredEvmScheme) EnhancePaymentRequirements(
	ctx context.Context,
	requirements types.PaymentRequirements,
	supportedKind types.SupportedKind,
	extensions []string,
) (types.PaymentRequirements, error) {
	return s.exact.EnhancePaymentRequirements(ctx, requirements, supportedKind, extensions)
}

File: typescript/x402-typescript/packages/mechanisms/evm/src/deferred/server/scheme.ts
import {
  AssetAmount,
  Network,
  PaymentRequirements,
  Price,
  SchemeNetworkServer,
} from "@okxweb3/app-x402-core/types";
import { getDefaultAsset, type ExactDefaultAssetInfo } from "../../shared/defaultAssets";

/**
 * EVM server implementation for the aggr_deferred payment scheme.
 *
 * Parses prices the same way as ExactEvmScheme. The OKX Facilitator batches
 * session key signatures on-chain asynchronously; the Seller receives
 * status=success immediately and delivers the resource without polling.
 */
export class AggrDeferredEvmScheme implements SchemeNetworkServer {
  readonly scheme = "aggr_deferred";

  async parsePrice(price: Price, network: Network): Promise<AssetAmount> {
    if (typeof price === "object" && price !== null && "amount" in price) {
      if (!price.asset) {
        throw new Error(`Asset address must be specified for AssetAmount on network ${network}`);
      }
      return {
        amount: price.amount,
        asset: price.asset,
        extra: price.extra || {},
      };
    }

    const amount = this.parseMoneyToDecimal(price);
    return this.defaultMoneyConversion(amount, network);
  }

  enhancePaymentRequirements(
    paymentRequirements: PaymentRequirements,
    supportedKind: {
      x402Version: number;
      scheme: string;
      network: Network;
      extra?: Record<string, unknown>;
    },
    extensionKeys: string[],
  ): Promise<PaymentRequirements> {
    void supportedKind;
    void extensionKeys;
    return Promise.resolve(paymentRequirements);
  }

  private parseMoneyToDecimal(money: string | number): number {
    if (typeof money === "number") {
      return money;
    }
    const cleanMoney = money.replace(/^\$/, "").trim();
    const amount = parseFloat(cleanMoney);
    if (isNaN(amount)) {
      throw new Error(`Invalid money format: ${money}`);
    }
    return amount;
  }

  private defaultMoneyConversion(amount: number, network: Network): AssetAmount {
    const assetInfo: ExactDefaultAssetInfo = getDefaultAsset(network);
    const tokenAmount = this.convertToTokenAmount(amount.toString(), assetInfo.decimals);
    const includeEip712Domain = !assetInfo.assetTransferMethod || assetInfo.supportsEip2612;
    return {
      amount: tokenAmount,
      asset: assetInfo.address,
      extra: {
        ...(includeEip712Domain && {
          name: assetInfo.name,
          version: assetInfo.version,
        }),
        ...(assetInfo.assetTransferMethod && {
          assetTransferMethod: assetInfo.assetTransferMethod,
        }),
      },
    };
  }

  private convertToTokenAmount(decimalAmount: string, decimals: number): string {
    const amount = parseFloat(decimalAmount);
    if (isNaN(amount)) {
      throw new Error(`Invalid amount: ${decimalAmount}`);
    }
    const [intPart, decPart = ""] = String(amount).split(".");
    const paddedDec = decPart.padEnd(decimals, "0").slice(0, decimals);
    const tokenAmount = (intPart + paddedDec).replace(/^0+/, "") || "0";
    return tokenAmount;
  }
}
